components/Dialogs/src/MDCDialogTransitionController.m

```diff
@@ -60,30 +71,60 @@ - (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionCo
   CGFloat endingAlpha = presenting ? 1 : 0;
 
   animatingView.frame = [transitionContext finalFrameForViewController:animatingViewController];
-  animatingView.alpha = startingAlpha;
 
-  NSTimeInterval transitionDuration = [self transitionDuration:transitionContext];
   UIViewAnimationOptions options =
       UIViewAnimationOptionAllowUserInteraction | UIViewAnimationOptionBeginFromCurrentState;
 
-  [UIView animateWithDuration:transitionDuration
-      delay:0.0
-      options:options
-      animations:^{
-        animatingView.alpha = endingAlpha;
-      }
-      completion:^(__unused BOOL finished) {
-        // If we're dismissing, remove the presented view from the hierarchy
-        if (!presenting) {
-          [fromView removeFromSuperview];
-        }
-
-        // From ADC : UIViewControllerContextTransitioning
-        // When you do create transition animations, always call the
-        // completeTransition: from an appropriate completion block to let UIKit know
-        // when all of your animations have finished.
-        [transitionContext completeTransition:YES];
-      }];
+  // Opacity animation
+  animatingView.alpha = startingAlpha;
+  [UIView animateWithDuration:_animationDuration / 2
+                        delay:0
+                      options:options
+                   animations:^{
+                     animatingView.alpha = endingAlpha;
+                   }
+                   completion:nil];
+
+  // Scale animation
+  MDCDialogPresentationController *presentationController = nil;
+  if ([animatingViewController.presentationController
+          isKindOfClass:[MDCDialogPresentationController class]]) {
+    presentationController =
+        (MDCDialogPresentationController *)animatingViewController.presentationController;
+  }
+
+  CGAffineTransform startingTransform =
+      presenting ? CGAffineTransformMakeScale(_dialogScaleFactor, _dialogScaleFactor)
+                 : CGAffineTransformIdentity;
+  CGAffineTransform endingTransform = CGAffineTransformIdentity;
+  animatingView.transform = startingTransform;
+  presentationController.dialogTransform = startingTransform;
+  UIViewAnimationOptions scaleAnimationOptions = options | UIViewAnimationOptionCurveEaseOut;
+  [UIView animateWithDuration:_animationDuration
+                        delay:0
+                      options:scaleAnimationOptions
+                   animations:^{
+                     animatingView.transform = endingTransform;
+                     presentationController.dialogTransform = endingTransform;
+                   }
+                   completion:nil];
+
+  // Use dispatch_after to avoid animateWithDuration immediately executing its completion block.
+  // This can happen if the animation has no effect (changing animatingView.transform from
+  // CGAffineTransformIdentity to CGAffineTransformIdentity, in this case).
+  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(_animationDuration * NSEC_PER_SEC)),
```

Comment:
what if we put all of the animation blocks surrounded by a CATransaction begin and commit, set the CATransaction duration to animationDuration and set its completion block to the completion below?

Reply:
I like that approach a lot better – the problem I was trying to balance here is that we have to completion blocks, and I didn't want to put a bunch of logic in both of them determining which should do the cleanup. Will fix